### PR TITLE
Forbid negative time snap in editor

### LINKED ISSUE #21944: Possible to place notes out of bound in taiko, catch and mania editor.
### Type

Game behaviour

### Bug description

Probably doesn't need explaination.

### Screenshots or videos

https://user-images.githubusercontent.com/11626920/210063744-34449712-e4f0-4bf4-b3fb-2b7a8e5df6db.mp4

### Version

2022.1228.0

### Logs

[database.log](https://github.com/ppy/osu/files/10324379/database.log)
[input.log](https://github.com/ppy/osu/files/10324380/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/10324381/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/10324382/network.log)
[performance.log](https://github.com/ppy/osu/files/10324383/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/10324384/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/10324385/performance-draw.log)
[runtime.log](https://github.com/ppy/osu/files/10324402/runtime.log)
[updater.log](https://github.com/ppy/osu/files/10324403/updater.log)
[performance-input.log](https://github.com/ppy/osu/files/10324386/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/10324388/performance-update.log)

### PR DESCRIPTION
**Pull request by Feodor0090**
Closes #21944 

Initially i wanted to somehow make it properly round to nearest positive beat, but stuck on dealing with edge cases so ended with `if(n<0) n+=x` thing to avoid overcomplicating.

### CODE REVIEW

File: osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs [hunk 1]
```diff
@@ -144,7 +144,7 @@ public void TestLengthAndStarRatingUpdated()
             double lastStarRating = 0;
             double lastLength = 0;
 
-            AddStep("Add timing point", () => EditorBeatmap.ControlPointInfo.Add(500, new TimingControlPoint()));
+            AddStep("Add timing point", () => EditorBeatmap.ControlPointInfo.Add(200, new TimingControlPoint { BeatLength = 600 }));
```
Comment (peppy) [inline osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs:147]: To save me going through the whole process, can you explain why this change was required?

File: osu.Game.Tests/NonVisual/ClosestBeatDivisorTest.cs [hunk 2]
```diff
@@ -17,7 +17,7 @@ public class ClosestBeatDivisorTest
         public void TestExactDivisors()
         {
             var cpi = new ControlPointInfo();
-            cpi.Add(-1000, new TimingControlPoint { BeatLength = 1000 });
+            cpi.Add(0, new TimingControlPoint { BeatLength = 1000 });
```
Comment (peppy) [inline osu.Game.Tests/NonVisual/ClosestBeatDivisorTest.cs:20]: Likewise, can you explain why the changes in this file were required?

File: osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs (hunk 1 above)
Comment (Feodor0090) [reply osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs:147]: The test is placing 2 circles. Before this PR first one was placed on negative time, so they started to stack on the same beat. I changed this to make them placed on different beats to actually contribute to map's length.

File: osu.Game.Tests/NonVisual/ClosestBeatDivisorTest.cs (hunk 2 above)
Comment (Feodor0090) [reply osu.Game.Tests/NonVisual/ClosestBeatDivisorTest.cs:20]: Test was working from point's start time so created objects on negative time. It can't work now (will round to 0 and upper) so i didn't spend too much time investigating this, just moved all operations to positive time.

**Review by peppy — APPROVED** [review]
Seems sane enough
